### PR TITLE
Handle exceptions thrown from stop hooks

### PR DESCRIPTION
**Pull request by TimMoore**
Related to https://github.com/playframework/playframework/pull/8497#discussion_r199424160

### CODE REVIEW

File: framework/src/play/src/test/scala/play/api/inject/DefaultApplicationLifecycleSpec.scala [hunk 1]
```diff
@@ -28,6 +28,28 @@ class DefaultApplicationLifecycleSpec extends Specification {
       Await.result(lifecycle.stop(), 10.seconds)
       buffer.toList must beEqualTo(List(3, 2, 1))
     }
+
+    "continue when a hook returns a failed future" in {
```
Comment (TimMoore) [inline framework/src/play/src/test/scala/play/api/inject/DefaultApplicationLifecycleSpec.scala]: This one already passed, but there wasn't an existing test for this behavior.

File: framework/src/play/src/test/scala/play/api/inject/DefaultApplicationLifecycleSpec.scala [hunk 2]
```diff
@@ -28,6 +28,28 @@ class DefaultApplicationLifecycleSpec extends Specification {
...
+      val buffer = mutable.ListBuffer[Int]()
+      lifecycle.addStopHook(() => Future(buffer.append(1)))
+      lifecycle.addStopHook(() => Future.failed(new RuntimeException("Failed stop hook")))
+      lifecycle.addStopHook(() => Future(buffer.append(3)))
+      Await.result(lifecycle.stop(), 10.seconds)
+      Await.result(lifecycle.stop(), 10.seconds)
+      buffer.toList must beEqualTo(List(3, 1))
+    }
+
+    "continue when a hook throws an exception" in {
```
Comment (TimMoore) [inline framework/src/play/src/test/scala/play/api/inject/DefaultApplicationLifecycleSpec.scala]: This tests the change to `DefaultApplicationLifecycle` above (and fails without it).

**Comment by TimMoore** [comment]
I just noticed that this will conflict with #8406, but I think the change is still desirable even in the context of those changes (CC: @ignasi35).

**Comment by marcospereira** [comment]
Backport to 2.6.x: 122629f326a5b6f2ff48f21dbba3f54fa27709b1

**Pull request merged**